gridstatus/tests/test_pjm.py

```diff
@@ -457,6 +526,109 @@ def _check_solar_forecast(self, df):
             skip_column_named_time=True,
         )
 
+    # TODO(kladar): Really what we want to probably test is that most importantly
+    # 0. The API is up and returns 200 and data (intergration test)
+    # 1. The data returned is in the expected format
+    # 1.1. The data types are correct (unit test)
+    # 1.2. The horizons are all represented (data test)
+    # 1.3. The interval is correct for the given resolution (data test)
+    # 1.4. Our transformations and renamings are correctly applied (unit test)
+    # 2. There are no data gaps or duplicates (data test)
+    @pytest.mark.integration
+    def test_get_hourly_solar_forecast_integration(self):
+        df = self.iso.get_solar_forecast("today")
+        assert isinstance(df, pd.DataFrame)
+        assert not df.empty
+        self._check_solar_forecast(df)
+
+    # NOTE(kladar): This is an example of a more pure integration test,
+    # it tests that the data is returned from the API and that the data is in the expected format,
+    # but does not check the comprehensiveness of the data, which is arguably a separate concern
+    @pytest.mark.integration
+    def test_get_5min_solar_forecast_integration(self):
+        df = self.iso.get_solar_forecast("today", resolution="5min")
+        assert isinstance(df, pd.DataFrame)
+        assert not df.empty
+        self._check_solar_forecast(df)
+
+    @pytest.mark.integration
+    def test_get_solar_forecast_historical_date_integration(self):
+        past_date = self.local_today() - pd.Timedelta(days=10)
+        df = self.iso.get_solar_forecast(past_date)
+        assert isinstance(df, pd.DataFrame)
+        assert not df.empty
+        self._check_solar_forecast(df)
+
+    @pytest.mark.integration
+    def test_get_solar_forecast_historical_range_integration(self):
+        past_date = self.local_today() - pd.Timedelta(days=12)
+        past_end_date = past_date + pd.Timedelta(days=3)
+        df = self.iso.get_solar_forecast(past_date, past_end_date)
+        assert isinstance(df, pd.DataFrame)
+        assert not df.empty
+        self._check_solar_forecast(df)
+
+    def test_get_solar_forecast_unit_5min(self, sample_5min_solar_forecast_data):
+        with mock.patch.object(
+            self.iso,
+            "get_solar_forecast",
+            return_value=sample_5min_solar_forecast_data,
+        ):
```

Comment:
It's not clear to me what this is testing. Since the function is mocked with the fixture, any changes to the function are not tested.

Reply:
Yeah true. It's poorly named for what it does now, since it was adapted from the prior integration-style test that did query the external service into simply not doing that. But this is more testing the fixture properties are what they say they are. Not super useful, but I guess not entirely useless. Mostly wanted this is to start introducing fixtures to test_pjm.py more generally and adapting the existing tests to them, with the assumption that they will evolve considerably from here.

We could remove it if we aren't interested in testing the fixture properties (but the test in only a few milliseconds).

Reply:
It would make more sense if we mocked `_get_pjm_json` with a fixture and checked that `get_solar_forecast` correctly handles the output. I think if we use mocking, we should get as close to the raw output from the external API as we can so our unit tests are checking the data parsing.

My main issue with only testing the fixture properties is it could give a false sense of security. If the source code changes but the developer doesn't update the fixture, the unit test will still pass even if the source code introduces an error.

Reply:
yes agreed, you read my mind on `_get_pjm_json()`, which is what I really want to unit test (as well as `_parse_solar_forecast` etc).  Was having some trouble getting at it yesterday and wanted to get the new forecast query logic out to you before too long and not spend any more time in the testing trenches, but next pass was going to be to get at `_get_pjm_json`. I can get there for this PR though, if we don't mind waiting a few more hours.

Reply:
I think we all our in agreement, but I'll just repeat it....we should fixtures as close as possible to the exact data that comes back from the source. anything else is embedding our transformation logic into the fixture. and our goal is to test the transformation logic. 

what I'll add is, maybe we should want to think about structuring each dataset's scraper a bit differently, so we can automatically reason more about each step/mock different steps

```python

class DatasetBase:
    def run(self):
        data = self.fetch_raw_data()
        data = self.process_raw_data(data)
        data = self.finalize_data(data)
        return data
    
class PJMBase(DatasetBase):
    """"Reusable pjm methods""""
    def query_pjm(self, url):
        pass

    def process_pjm_response(self, response):
        pass

class PJMSolarForecastHourly(PJMBase):
    def fetch_raw_data(self):
        """returns data as raw as possible from source ie json, file, etc"""
        return self.query_pjm(...)

    def process_raw_data(self, data):
        """standardized processing of raw data across many datasets"""
        return self.process_pjm_response(data)
        
    def finalize_data(self, data):
        """dataset specific processing"""
        return data
```

I dont know if this is right approach but just want to through something out there


practically speaking, i think this is a bigger discussion, so no need to implement too much yet. i think we merge this dataset in and then set up time to discuss the ideal end state across repo

Reply:
Yeah I agree with all of the above, and gave it a day of trying to test things in a way that worked with the existing structure in case it was straightforward and extensible across the repo, and struggled to come up with anything that was compelling. Found it hard to create fixtures for the _general_ methods (`_get_json`, `_get_pjm_json`) that can then be reused in the _specific_ methods (`get_solar_forecast_5_minute`) without heavily duplicating/repeating the build up methods or heavily parameterizing the fixtures and having conditional logic to the specific method tests. 

Tried 3-4 approaches and didn't find anything I'd put my stamp on. I built up some new tests for the general methods with the fixtures and left it at that, and we can revisit.

Reply:
Captured this in ETL v3, resolving.